### ISSUE TITLE
`include()` runs code before discovering trailing syntax errors

### ISSUE THREAD

**Issue by c42f** [JuliaLang/julia#44610]
As noted [on discourse](https://discourse.julialang.org/t/a-most-harrowing-collection-of-julia-wats/77323/50?u=chris_foster), `include()` executes top-level code incrementally, even when that code has known trailing syntax errors. As a reduced example:

```julia
println("What a lovely day!")

# ... <-- Some potentially long-running code

println("WAT!"));  # A syntax error
```

This isn't a matter of incremental parsing.  The expression which is being executed here is parsed eagerly into an `Expr(:toplevel)` as

```julia
:($(Expr(:toplevel,
         :(println("What a lovely day!")),
         :($(Expr(:error, "extra token \")\" after end of expression"))))))
```

So it would be simple to detect the trailing syntax error immediately and refuse to run the code as a whole, rather than executing a bunch of potentially side-effectful code.  Note that the REPL already has this behavior — it uses the logic in the oddly-named `Base.parse_input_line()` to achieve this.

Changing `include()` like this would be technically breaking. But generally it seems like a good quality of life improvement.

**Comment by c42f** [comment]
> Does it?

Yes.

> If I paste that code above into the REPL, the first println runs.

You can't copy-paste it in directly to simulate this — if you do that the REPL will execute after the first newline where it detects the first complete Julia expression.

But you can use ALT+enter to enter newlines without asking for the REPL to execute them.  So you need to type something like

`printlin("hi")` ALT+enter `println("ho"))` enter

**Comment by oscarrutt** [comment]
Hi. I posted https://discourse.julialang.org/t/a-most-harrowing-collection-of-julia-wats/77323/50?u=chris_foster and thanks to @c42f  for moving it here.

To objections that this behavior isn't a problem or is in the spirit of Julia, my point is this: Except for small problems and one-off plots, I can't seriously use a language which runs a simulation for an hour or a few hours or even 10 minutes then looses all the results because the entire piece of code wasn't checked for syntax errors at the beginning.

**Comment by heliosdrm** [comment]
I agree that it is good to fail as soon as possible, but the effectiveness of the proposed solution depends on how would it behave with nested `include`s: if a file that would fail early for having such syntax errors does not trigger a similar error in a file that `include`s it, that wouldn't be so useful.

**Comment by c42f** [comment]
> if a file that would fail early for having such syntax errors does not trigger a similar error in a file that includes it, that wouldn't be so useful

It's not possible to detect syntax errors in nested includes because `include()` is dynamic: it takes a path which may be computed at runtime.

But even without this I think it'd be useful to emit syntax errors eagerly — it would just be a good quality of life improvement for the average scripting use case.
